GVRf/Framework/framework/src/main/jni/objects/components/render_data.cpp

```diff
@@ -149,161 +112,20 @@ void RenderData::bindShader(JNIEnv* env, jobject localSceneObject, bool isMultiv
-            && i->rendering_order() < RenderData::Overlay)
-        {
-            return i->camera_distance() > j->camera_distance();
-        }
-
-        if (i->get_shader(0) == j->get_shader(0))
-        {
-            int no_passes1 = i->pass_count();
-            int no_passes2 = j->pass_count();
-
-            if (no_passes1 == no_passes2)
-            {
-                //@todo what about the other passes
-
-                //this is pointer comparison; assumes batching is on; if the materials are not
-                //the same then comparing the pointers further is an arbitrary decision; hence
-                //falling back to camera distance.
-                if (i->material(0) == j->material(0))
-                {
-                    if (i->cull_face(0) == j->cull_face(0))
-                    {
-                        if (i->getHashCode().compare(j->getHashCode()) == 0)
-                        {
-                            // otherwise sort from front to back
-                            return i->camera_distance() < j->camera_distance();
-                        }
-                        return i->getHashCode() < j->getHashCode();
-                    }
-                    return i->cull_face(0) < j->cull_face(0);
-                }
-                return i->material(0) < j->material(0);
-            }
-            return no_passes1 < no_passes2;
-        }
-        return i->get_shader() < j->get_shader();
-    }
-    return i->rendering_order() < j->rendering_order();
-}
-
 const std::string& RenderData::getHashCode()
 {
-    if (hash_code_dirty_)    {
+    if (isHashCodeDirty())
+    {
         std::string render_data_string;
         render_data_string.append(std::to_string(getRenderDataFlagsHashCode()));
         render_data_string.append(std::to_string(getComponentType()));
-        render_data_string.append(std::to_string(render_mask_));
-        render_data_string.append(std::to_string(offset_factor_));
-        render_data_string.append(std::to_string(offset_units_));
-        render_data_string.append(std::to_string(sample_coverage_));
-        render_data_string.append(std::to_string(stencilMaskMask_));
-        render_data_string.append(std::to_string(stencilFuncFunc_));
-        render_data_string.append(std::to_string(stencilFuncRef_));
-        render_data_string.append(std::to_string(stencilFuncMask_));
-        render_data_string.append(std::to_string(stencilOpSfail_));
-        render_data_string.append(std::to_string(stencilOpDpfail_));
-        render_data_string.append(std::to_string(stencilOpDppass_));
-        render_data_string.append(std::to_string(source_alpha_blend_func_));
-        render_data_string.append(std::to_string(dest_alpha_blend_func_));
         render_data_string.append(std::to_string(mesh_->getVertexBuffer()->getDescriptor()));
         hash_code = render_data_string;
-        hash_code_dirty_ = false;
+        pass(0)->render_modes().clearDirty();
```

Comment:
not for all the passes?

Reply:
We have to talk about the hash code. All of the render states now fit in a single 64 bit word, one for each pass. The render passes are rendered separately, not together anymore. OpenGL doesn't use this hash code. We should discuss its significance in Vulkan.